assembl/auth/social_auth.py

```diff
@@ -218,3 +236,53 @@ def get_pipeline(self):
             # Update the user record with any changed info from the auth service.
             'assembl.auth.social_auth.user_details'
         )
+
+
+def get_active_auth_strategies(settings):
+    """Give the list of available social auth providers.
+    Includes multiple instances if a provider can have multiple servers.
+    This currently includes SAML, and eventually wordpress.
+    TODO: Should replace the login_providers config variable"""
+    all_backends = load_backends(settings.get('SOCIAL_AUTH_AUTHENTICATION_BACKENDS'))
+    for backend_name in all_backends:
+        def get_setting(name):
+            return (settings.get(setting_name(SETTING_PREFIX, backend_name, name), None)
+                    or settings.get(setting_name(backend_name, name), None))
+        if backend_name == 'wordpress-oauth2':
+            # TODO: This special case needs to be treated the same as saml asap.
+            # Also: maybe check preferences
+            yield backend_name
+        elif backend_name == 'saml':
+            # special case: Multiple IDPs
+            idps = get_setting('ENABLED_IDPS') or {}
+            for idp in idps.keys():
+                yield 'saml:' + idp
+        elif get_setting('key'):
+            yield backend_name
+
+
+def includeme(config):
+    """Pre-parse certain settings for python_social_auth, then load it."""
+    settings = config.get_settings()
+    settings['login_providers'] = aslist(settings.get('login_providers', ''))
+    settings['trusted_login_providers'] = aslist(settings.get('trusted_login_providers', ''))
+    if not any(settings['login_providers']):
+        sys.stderr.write('no login providers configured, double check '
+                         'your ini file and add a few')
+    for name in ('SOCIAL_AUTH_AUTHENTICATION_BACKENDS',
+                 'SOCIAL_AUTH_USER_FIELDS',
+                 'SOCIAL_AUTH_PROTECTED_USER_FIELDS',
+                 'SOCIAL_AUTH_FIELDS_STORED_IN_SESSION'):
+        settings[name] = aslist(settings.get(name, ''))
+    for name in ('SOCIAL_AUTH_SAML_ORG_INFO',
+                 'SOCIAL_AUTH_SAML_TECHNICAL_CONTACT',
+                 'SOCIAL_AUTH_SAML_SUPPORT_CONTACT',
+                 'SOCIAL_AUTH_SAML_ENABLED_IDPS'):
+        val = settings.get(name, '')
+        if val:
+            settings[name] = json.loads(val)
+    for k in settings.iterkeys():
+        if k.endswith("_SCOPE") and k.startswith("SOCIAL_AUTH_"):
+            settings[k] = aslist(settings.get(k, ''))
```

Comment:
pyramid-social-auth expects lists/json in those preferences, not strings. I'm preparing the field for it.